docs/helper-environment-visitor.md

```diff
@@ -0,0 +1,285 @@
+  "key": { "type": "AssignmentExpression" }, // [this.foo = "outer"]
+  "body": { "type": "BlockStatement" }, // { this.foo = "inner"; }
+  "params": [], // should visit too if there are any
+  "computed": true
+}
+```
+If the entire `ClassMethod` node is skipped, then we won't be able to visit the `this.foo` under the `key` property. However, we must visit it as it could be any expression. To achieve this, we need to tell Babel to skip only the `ClassMethod` node, but _not_ its computed key. This is where [`requeueComputedKeyAndDecorators`](#requeuecomputedkeyanddecorators) comes in handy:
+
+```js title="Revision 4: replace-top-level-this-plugin.js"
+import {
+  requeueComputedKeyAndDecorators
+} from "@babel/helper-environment-visitor";
+
+module.exports = (api) => {
+  const { types: t } = api;
+  return {
+    name: "replace-top-level-this",
+    visitor: {
+      ThisExpression(path) {
+        path.replaceWith(t.identifier("globalThis"));
+      }
+      FunctionParent(path) {
+        if (!path.isArrowFunctionExpression()) {
+          path.skip();
+        }
+        // highlight-start
+        if (path.isMethod()) {
+          requeueComputedKeyAndDecorators(path);
+        }
+        // highlight-end
+      }
+    }
+  }
+}
+```
+
+There is still one missing edge case: `this` can be in computed keys of a class property:
+
+```js title="input.js"
+class Bar {
+  // replace
+  [this.foo = "outer"] =
+  // don't replace
+  this.foo
+}
+```
+
+Although `requeueComputedKeyAndDecorators` can handle this edge case as well, the plugin has become quite complex at this point, with a significant amount of time spent on handling the `this` context. In fact, the focus on dealing with `this` has detracted from the actual requirement, which is to replace top-level `this` with `globalThis`.
+
+The `environmentVisitor` is created to simplify the code by extracting the error-prone `this`-handling logic into a helper function, so that you no longer have to deal with it directly.
+
+```js title="Revision 5: replace-top-level-this-plugin.js"
+import environmentVisitor from "@babel/helper-environment-visitor";
+
+module.exports = (api) => {
+  // highlight-next-line
+  const { types: t, traverse } = api;
+  return {
+    name: "replace-top-level-this",
+    // highlight-next-line
+    visitor: traverse.visitors.merge([
+      {
+        ThisExpression(path) {
+          path.replaceWith(t.identifier("globalThis"));
+        }
+      },
+      // highlight-start
+      environmentVisitor
+    ]);
+    // highlight-end
+  }
+}
+```
+
+You can try out the final revision on the [AST Explorer](https://astexplorer.net/#/gist/4d1bfca2b315f687da44f3436b2f4d76/58c9b92d4f77586f23f56393252104a274ccb157).
+
+As you may find out from the name, `requeueComputedKeyAndDecorators` supports [ES decorators](./plugin-proposal-decorators.md), too:
+
+```js title="input.js"
+class Foo {
+  // replaced to `@globalThis.log`
+  @this.log foo = 1;
```

Comment:
```suggestion
  @(this.log) foo = 1;
```
The current `DecoratorMemberExpression` production does not allow `this.log` (https://github.com/tc39/proposal-decorators/issues/460).